File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        self.lastv = None
        
        def dfs(n):
            if not n:
                return True
            
            if not dfs(n.left):
                return False
            
            if self.lastv!=None and self.lastv>=n.val:
                return False
            self.lastv = n.val
            
            if not dfs(n.right):
                return False
            
            return True
            
        return dfs(root)

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if(!root)
            return true;
        
        if(invalleft(root->left, root) || invalright(root->right, root))
            return false;
        
        return isValidBST(root->left) && isValidBST(root->right);
    }
    
    int invalleft(TreeNode *node, TreeNode *root){
        if(!node)
            return false;
        while(node->right)
            node=node->right;
        return node->val >= root->val;
    }
    
    int invalright(TreeNode *node, TreeNode *root){
        if(!node)
            return false;
        while(node->left)
            node=node->left;
        return node->val <= root->val;
    }
};